tests/integration/conftest.py

```diff
@@ -191,3 +201,161 @@ async def kubernetes_cluster(request: pytest.FixtureRequest, ops_test: OpsTest):
         bundle.switch(charm.app_name, path)
     async with deploy_model(request, ops_test, model, bundle) as the_model:
         yield the_model
+
+
+@pytest_asyncio.fixture(scope="module")
+async def grafana_agent(ops_test: OpsTest, kubernetes_cluster: Model):
+    """Deploy Grafana Agent."""
+    with ops_test.model_context("main") as model:
+        await model.deploy("grafana-agent", channel="stable")
+        await model.integrate("grafana-agent:cos-agent", "k8s:cos-agent")
+        await model.integrate("grafana-agent:cos-agent", "k8s-worker:cos-agent")
+        await model.integrate("k8s:cos-worker-tokens", "k8s-worker:cos-tokens")
+
+        yield
+
+        await model.remove_application("grafana-agent")
+
+
+@pytest_asyncio.fixture(scope="module")
+async def cos_model(ops_test: OpsTest, kubernetes_cluster: Model, grafana_agent):
+    """Create a COS substrate and a K8s model."""
+    log.info("Creating COS substrate...")
+    container_name = "cos-substrate"
+    network_name = "cos-network"
+    manager = LXDSubstrate(container_name, network_name)
```

Comment:
hold up... are you saying that to test cos -- you must test with microk8s on lxd?  there's no other option?

Reply:
Yes, I attempted to run the COS stack on CK. However, the charms are not coming up due to some issues with applying container limits. I strongly believe this is the way to test COS because we are using an external machine where our COS stack is running (probably on another subnet as well).

Reply:
what about COS on uK8s -- just running from a machine deployed via juju rather than pylxd? 

some developers may wish to use their cloud resources to test rather than the the local machine where the tests run from. 

#team-inspiron